### ISSUE TITLE
feature req: CI workflow without matrix to minimize GHA runtime

### ISSUE THREAD

**Issue by thenonameguy** [nixbuild/nixbuild-action#28]
The UX of the CI workflow is great. The only downside is that private Github repos have a limited free tier for GHA runner time.
Currently if 10 jobs execute in the matrix in parallel for 2 minutes on nixbuild.net, this consumes 20 minutes of that time.
The only gain from this attr-per-job matrix model AFAIK is isolated logs / per-job error messages in case of failures.

The request is to get the above scenario to consume 2 minutes of GHA time.
This means logs would have to be streamed from the remote builders (ala: https://github.com/maralorn/nix-output-monitor) but in a CI context.
A quick and dirty implementation might to create a  [linkFarm](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders.nix#L488) for the `jq` selected Flake attributes and building it `--keep-going` depending on `fail-fast`.

**Comment by rickynils** [comment]
This is very valid feedback. While the usage of nixbuild.net is really streamlined (due to build "deduplication"), the GH Actions build time is burned quickly with the current setup.

The initial reason for the build matrix was to improve Nix evaluation performance a bit by parallelizing it over multiple runners, but mostly to get a nicer UI where you can see passed/failed attributes easily, and jump to failures quickly.

Since GitHub now has a nicer way of generating workflow summaries in markdown, we could perhaps do all evaluation and remote building on a single runner, and then compile the summary in the end. We do this today too, but we could make the summary more attribute-centered instead (or in addition to) the current build-centered view.

What we would lose is the incremental feedback of passing/failing attributes, and the interactive UI (perhaps somewhat reproducible in the markdown summary).

**Comment by rickynils** [comment]
I've also been thinking if it would somehow be possible to run _asynchronous_ jobs. Like, we could do the evaluation like we do today, kick off the build and then quit the GHA job. Then somehow, when the build is done, nixbuild.net would call back into GHA. Something like that is described briefly [here](https://www.youtube.com/watch?v=-xrFNFby7hc), using _repository dispatch_. However, I haven't looked into how it would work in practice, maybe it just becomes very messy.

**Comment by thenonameguy** [comment]
Hah, I've checked out the `repository_dispatch` CVE example on Github before opening this issue. 

The async job solution suffers from not having much introspection capabilities:
- no logs without extra work
- no way of checking the build progress outside of the current `nixbuild.net` ssh interface

[This approach sounds great to me](https://github.com/nixbuild/nixbuild-action/issues/28#issuecomment-1342781813).
The 10x reduction in the GHA bill can be spent on nixbuild.net, which is the primary goal for me.
The async last-mile is out of scope for my own needs. 

Just to bring up one more alternative:
Nixbuild.net could act as a 'self-hosted runner' by implementing the necessary webhooks.
That would be probably the cleanest implementation, eliminating the GHA runner completely.
Something like: https://buildjet.com/for-github-actions

**Comment by rickynils** [comment]
> Just to bring up one more alternative:
Nixbuild.net could act as a 'self-hosted runner' by implementing the necessary webhooks.
That would be probably the cleanest implementation, eliminating the GHA runner completely.
Something like: https://buildjet.com/for-github-actions

Do you know how they do that? Are they running the ordinary self-hosted github runner on their servers, or doing something more magical? You mention webhooks, you have any pointers to that? This would indeed be a clean solution, but also more maintenance that would be specific to GitHub.

**Comment by thenonameguy** [comment]
I haven't ever done it, but these are open-source implementation for AWS clouds:
https://github.com/machulav/ec2-github-runner (slimmer, relatively unknown)
https://github.com/philips-labs/terraform-aws-github-runner (more popular, from the top household appliance company)

Should be portable quite easily to other clouds (especially bare-metal/KVM ones that you are supposedly using).

General docs: https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners

**Comment by rickynils** [comment]
@thenonameguy Thanks for the hints about the self-hosted runner implementations, I'll look into that.

I had another idea: We could add an option to `ci-workflow` for configuring attributes that should be built before the matrix. So you could build packages that you know will be needed by all the matrix jobs first, before firing up all concurrent jobs. Of course, you're then doing some of the depedency resolution manually which seems stupid when you have Nix, but at least it feels like a solution that can be implemented quickly :)

**Comment by thenonameguy** [comment]
@rickynils any way I can help pushing this forward? Would you be open to a PR?

**Comment by rickynils** [comment]
@thenonameguy A PR would be much welcome!

**Comment by KaiserKarel** [comment]
Alternatively, we could spawn a single job, which uses the matrix to create extra checks using GitHub script, that way we do not pay GA minutes for every spawned job, just for one orchestrator, and still get the same UI elements as the current job has.

https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#create-a-check-run
